pkg/bloomcompactor/utils.go

```diff
@@ -0,0 +1,37 @@
+package bloomcompactor
+
+import "github.com/grafana/loki/pkg/storage/stores/shipper/bloomshipper"
+
+func matchingBlocks(metas []bloomshipper.Meta, job Job) ([]bloomshipper.Meta, []bloomshipper.BlockRef) {
+	var metasMatchingJob []bloomshipper.Meta
+	var blocksMatchingJob []bloomshipper.BlockRef
+	var oldTombstonedBlockRefs map[bloomshipper.BlockRef]struct{}
+
+	for _, meta := range metas {
+		if meta.TableName != job.tableName {
+			continue
+		}
+		metasMatchingJob = append(metasMatchingJob, meta)
+
+		for _, tombstonedBlockRef := range meta.Tombstones {
+			oldTombstonedBlockRefs[tombstonedBlockRef] = struct{}{}
+		}
+	}
+
+	for _, meta := range metasMatchingJob {
+		for _, blockRef := range meta.Blocks {
+			if _, ok := oldTombstonedBlockRefs[blockRef]; ok {
+				// skip any previously tombstoned blockRefs
+				continue
+			}
+
+			if blockRef.IndexPath == job.indexPath {
+				// index has not changed, no compaction needed
+				continue
+			}
+			blocksMatchingJob = append(blocksMatchingJob, blockRef)
+		}
+	}
```

Comment:
The idea of two loops is to accumulate all previously tombstoned refs in one go first. if a block is later seen to in tombstoned list then i will have to remove it from `blocksMatchingJobs`. 

Instead of removing from a list, i think this is cleaner to read. block lists should not grow super big, so i didn't think it's a deal breaker. Please let me know what you prefer here though.